File: 1557.py
while True:
    n = int(input())
    if n == 0:
        break

    # calcula largura do maior número
    maior = 2 ** (2 * (n - 1))
    width = 0
    temp = maior
    while temp > 0:
        temp //= 10
        width += 1

    count = 0
    for i in range(n):
        a = 0
        linha = ""
        for j in range(n):
            aux = 2 ** count
            if j == 0:
                linha += f"{aux:>{width}}"
            else:
                linha += f" {aux:>{width}}"
            count += 1
            a += 1
        count = count - (a - 1)
        print(linha)
    print()

File: 1557.c
#include <math.h>
#include <stdio.h>

int main() {
    int n, i, j, aux, count, a, width, maior;

    while(1) {
        scanf("%d", &n);
        if(n == 0)
            break;
        
        maior = (int) pow(2, 2*(n-1));
        width = 0;
        int temp = maior;
        while(temp > 0) {
            temp /= 10;
            width++;
        }

        count = 0;
        for(i = 0; i < n; i++) {
            a = 0;
            for(j = 0; j < n; j++) {
                aux = (int) pow(2, count);
                if(j == 0)
                    printf("%*d", width, aux);
                else
                    printf(" %*d", width, aux);
                count++;
                a++;
            }
            count = count - (a - 1);
            printf("\n");
        }
        printf("\n");
    }

    return 0;
}
